modules/concurrency/include/hephaestus/concurrency/when_all_range.h

```diff
@@ -0,0 +1,175 @@
+//=================================================================================================
+// Copyright (C) 2023-2025 HEPHAESTUS Contributors
+//=================================================================================================
+
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <exception>
+#include <memory>
+#include <optional>
+#include <ranges>
+#include <type_traits>
+#include <utility>
+
+#include <stdexec/__detail/__execution_fwd.hpp>
+#include <stdexec/__detail/__senders_core.hpp>
+#include <stdexec/execution.hpp>
+
+namespace heph::concurrency {
+template <typename Range>
+concept SenderRange = std::ranges::range<Range> && stdexec::sender<std::ranges::range_value_t<Range>>;
+
+namespace internal {
+
+struct WhenAllStopCallback {
+  stdexec::inplace_stop_source* stop_source;
+  void operator()() const {
+    stop_source->request_stop();
+  }
+};
+
+enum WhenAllRangeState : std::uint8_t { STARTED, ERROR, STOPPED };
+
+template <typename Receiver>
+struct WhenAllRangeStateT {
+  using ReceiverEnv = stdexec::env_of_t<Receiver>;
+  using StopToken = stdexec::stop_token_of_t<ReceiverEnv>;
+  using StopCallbackT = stdexec::stop_callback_for_t<StopToken, WhenAllStopCallback>;
+
+  void arrive() noexcept {
+    if (1 == count.fetch_sub(1, std::memory_order_release)) {
```

Comment:
Using `release` here is suspicious as we are reading from `count`. Did you mean to use `acquire`? Or if you did want release semantics here, I think we also need acquire semantics, so `std::memory_order::acq_rel` is probably the safest bet.